common/src/gas_provider/internal.rs

```diff
@@ -629,6 +640,133 @@ where
         Self::create_from_with_value(key, new_key, amount, true)
     }
 
+    fn lock(key: Self::Key, amount: Self::Balance) -> Result<(), Self::Error> {
+        // Taking node to lock into.
+        let node = Self::get_node(key).ok_or_else(InternalError::node_not_found)?;
+
+        // Validating node type to be able to lock.
+        if node.is_reserved_local() {
+            return Err(InternalError::forbidden().into());
+        }
+
+        // Validating that node is not consumed.
+        if node.is_consumed() {
+            return Err(InternalError::node_was_consumed().into());
+        }
+
+        // Quick quit on queried zero lock.
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        // Taking value provider for this node.
+        let (mut upstream_node, node_id) = Self::node_with_value(node)?;
+
+        // Mutating value of provider.
+        let upstream_node_value = upstream_node
+            .value_mut()
+            .ok_or_else(InternalError::unexpected_node_type)?;
+
+        if *upstream_node_value < amount {
+            return Err(InternalError::insufficient_balance().into());
+        }
+
+        *upstream_node_value = upstream_node_value.saturating_sub(amount);
+
+        // If provider is a parent, we save it to storage, otherwise mutating
+        // current node further, saving it afterward.
+        let mut node = if let Some(upstream_id) = node_id {
+            StorageMap::insert(upstream_id, upstream_node);
+
+            Self::get_node(key).ok_or_else(InternalError::node_not_found)?
```

Comment:
unreachable, right?

Reply:
Sure. Queried the same above. Clone not used just to avoid excess cloning.

Reply:
Expects and any execution-breakers should be called on higher layers

I can just comment here, that this is unreachable error